src/lock.go

```diff
@@ -0,0 +1,95 @@
+//
+// Copyright (c) 2016-2017 Snowplow Analytics Ltd. All rights reserved.
+//
+// This program is licensed to you under the Apache License Version 2.0,
+// and you may not use this file except in compliance with the Apache License Version 2.0.
+// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the Apache License Version 2.0 is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+//
+
+package main
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/nightlyone/lockfile"
+)
+
+// Lock interface abstracting over file-based or consul-based locks
+type Lock interface {
+	TryLock() error
+	Unlock() error
+}
+
+// FileLock is for file-based locks
+type FileLock struct {
+	lock lockfile.Lockfile
+}
+
+// InitFileLock builds a FileLock (a temp file) with the argument as name
+func InitFileLock(name string) (Lock, error) {
+	lock, err := lockfile.New(filepath.Join(os.TempDir(), name))
```

Comment:
I don't think a temp dir is a safe place for a lock. Rather we should assume that the supplied lock path is an absolute path.